### PR TITLE
plugins/bf4: fix JSON encoding.

### PR DESCRIPTION
**Pull request by mkrautz**
The previous code accidently used the
C-string "host" and "serverID" variables
in the JSON output for the context and
identity.

To avoid too many changes, this commit
changes the JSON escape function to
work on C strings instead, and uses
that to escape the "host" and "serverID"
strings from the game before emitting
them into the JSON documents.

This commit also adds some extra code
to ensure that the game strings we
include in the JSON document are
within the printable ASCII range.

### CODE REVIEW

**Comment by mkrautz** [comment]
Note: not build tested.

**Comment by mkrautz** [comment]
PTAL

@davidebeatrici

File: plugins/bf4/bf4.cpp [hunk 1]
```diff
@@ -70,25 +100,22 @@ static int fetch(float *avatar_pos, float *avatar_front, float *avatar_top, floa
...
     context = ocontext.str();
 
     std::wostringstream oidentity;
     oidentity << "{";
     host[sizeof(host)-1] = 0; // NUL terminate queried C strings. We do this to ensure the strings from the game are NUL terminated. They should be already, but we can't take any chances.
-    std::string Host(host);
-    Host = jsonStringEscape(Host);
-    if (Host.find("bot") == std::string::npos) // Set host string as empty if "bot" is found in it.
+    escape(host);
+    if (strstr(host, "bot") == NULL) // Set host string as empty if "bot" is found in it.
```
Comment (mkrautz) [inline plugins/bf4/bf4.cpp]: This comment isn't correct. It wasn't before either.

What's the idea, @davidebeatrici?

File: plugins/bf4/bf4.cpp [hunk 2]
```diff
@@ -70,25 +100,22 @@ static int fetch(float *avatar_pos, float *avatar_front, float *avatar_top, floa
...
         if (!Host.empty()) {
```
Comment (mkrautz) [inline plugins/bf4/bf4.cpp]: Host doesn't exist anymore.

File: plugins/bf4/bf4.cpp (hunk 1 above)
Comment (mkrautz) [inline plugins/bf4/bf4.cpp]: Talked to @davidebeatrici on IRC -- the intention is what the code does: if "host" contains "bot", the JSON should not contain a Host.

File: plugins/bf4/bf4.cpp [hunk 3]
```diff
@@ -70,25 +100,22 @@ static int fetch(float *avatar_pos, float *avatar_front, float *avatar_top, floa
     }
 
     serverid[sizeof(serverid)-1] = 0; // NUL terminate queried C strings. We do this to ensure the strings from the game are NUL terminated. They should be already, but we can't take any chances.
-    std::string ServerID(serverid);
-    ServerID = jsonStringEscape(ServerID); // Replace '"' character with ' ' (space) to prevent JSON break.
+    escape(serverid);
     std::ostringstream ocontext;
-    ocontext << " {\"Server ID\": \"" << ServerID << "\"}"; // Set context with server ID
+    ocontext << " {\"Server ID\": \"" << serverID << "\"}"; // Set context with server ID
```
Comment (davidebeatrici) [inline plugins/bf4/bf4.cpp]: This needs to be changed in "serverid".

**Comment by davidebeatrici** [comment]
The plugin is working flawlessly.
I changed **host** and team **system**: https://www.diffchecker.com/af9ixyoq

**Comment by mkrautz** [comment]
Landed via bed423eaa168d8102a7ca3337e935128876ce776.

**Pull request merged**